src/object-store/src/layers/lru_cache/read_cache.rs

```diff
@@ -0,0 +1,290 @@
+    file_cache: Arc<C>,
+    /// Local memory cache to track local cache files
+    mem_cache: Cache<String, ReadResult>,
+}
+
+impl<C: Accessor + Clone> ReadCache<C> {
+    /// Create a [`ReadCache`] with capacity in bytes.
+    pub(crate) fn new(file_cache: Arc<C>, capacity: usize) -> Self {
+        let file_cache_cloned = file_cache.clone();
+        let eviction_listener =
+            move |read_key: Arc<String>, read_result: ReadResult, cause| -> ListenerFuture {
+                // Delete the file from local file cache when it's purged from mem_cache.
+                decrement_gauge!(OBJECT_STORE_LRU_CACHE_ENTRIES, 1.0);
+                let file_cache_cloned = file_cache_cloned.clone();
+
+                async move {
+                    if let ReadResult::Success(size) = read_result {
+                        decrement_gauge!(OBJECT_STORE_LRU_CACHE_BYTES, size as f64);
+
+                        let result = file_cache_cloned.delete(&read_key, OpDelete::new()).await;
+                        debug!(
+                            "Deleted local cache file `{}`, result: {:?}, cause: {:?}.",
+                            read_key, result, cause
+                        );
+                    }
+                }
+                .boxed()
+            };
+
+        Self {
+            file_cache,
+            mem_cache: Cache::builder()
+                .max_capacity(capacity as u64)
+                .weigher(|_key, value: &ReadResult| -> u32 {
+                    // TODO(dennis): add key's length to weight?
+                    value.size_bytes()
+                })
+                .async_eviction_listener(eviction_listener)
+                .support_invalidation_closures()
+                .build(),
+        }
+    }
+
+    /// Returns the cache's entry count and total approximate entry size in bytes.
+    pub(crate) async fn stat(&self) -> (u64, u64) {
+        self.mem_cache.run_pending_tasks().await;
+
+        (self.mem_cache.entry_count(), self.mem_cache.weighted_size())
+    }
+
+    /// Invalidte all cache items which key starts with `prefix`.
+    pub(crate) async fn invalidate_entries_with_prefix(&self, prefix: String) {
+        // Safety: always ok when building cache with `support_invalidation_closures`.
+        self.mem_cache
+            .invalidate_entries_if(move |k: &String, &_v| k.starts_with(&prefix))
+            .ok();
+    }
+
+    /// Blocking version of `invalidate_entries_with_prefix`.
+    pub(crate) fn blocking_invalidate_entries_with_prefix(&self, prefix: String) {
+        // Safety: always ok when building cache with `support_invalidation_closures`.
+        self.mem_cache
+            .invalidate_entries_if(move |k: &String, &_v| k.starts_with(&prefix))
+            .ok();
+    }
+
+    /// Recover existing cache items from `file_cache` to `mem_cache`.
+    /// Return entry count and total approximate entry size in bytes.
+    pub(crate) async fn recover_cache(&self) -> Result<(u64, u64)> {
```

Comment:
Can we fully trust the freshness of files in `file_cache`?

Consider the following situation: first write success (files created in both remote & local cache) -> update but crash happens (remote is written successfully but not for the local cache, resulting in inconsistency) -> recover cache from local after restarting

Reply:
It configures the file cache to use rename to ensure the integrity
https://github.com/GreptimeTeam/greptimedb/pull/2522#discussion_r1349472302

Reply:
https://github.com/GreptimeTeam/greptimedb/blob/0593c3bde389548931209eb1aa658daf2824618f/src/object-store/src/layers/lru_cache.rs#L95-L100

What I mean is not the atomicity of file operations. I mean that the crash may occur in L97. At this time, the old cache has not been deleted, and the results are inconsistent.

Reply:
Yes, that's possible in theory. But in our use case,  the files(except the manifest last checkpoint metadata) are all append-only and immutable, and we have already added the last checkpoint metadata file to the forbidden cache list. So this case should not happen in our scenario.

Reply:
I see

Reply:
It's tricky but works, we can find a better solution in the future.